### PR TITLE
[Taras Yakymiuk]Slideshow

### PR DESCRIPTION
**Pull request by taras1313**

### CODE REVIEW

File: src/app/home-page/components/slideshow/slideshow.component.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<div class="slidesWindow" (click)="stopSlideshow()" (mouseleave)='continueSlideshow()' (mouseenter)='pauseSlideshow()'>
```
[the remark this answers is not in the document]
Comment (dverbovyi) [reply src/app/home-page/components/slideshow/slideshow.component.html]: this template relates to slideshow component.

the method names called _continue_ and _pause_ are enough. Omit Slideshow from naming

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+@Component({
+  selector: 'app-slideshow',
+  templateUrl: './slideshow.component.html',
+  styleUrls: ['./slideshow.component.scss']
+})
+
+export class SlideshowComponent implements OnInit, AfterViewInit, OnDestroy {
+  @ViewChild('allSlides') allSlides: ElementRef;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: rename allSlides to slidesHolder as className . It's more appropriate name

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  private translateStep: number;
+  private currentTranslatePosition: number;
+  private totalSlidesSize: number;
+  private firstItem: HTMLElement;
+  private lastItem: HTMLElement;
+  private counter = 0;
+  public selected = 0;
+  private intervalStart;
+  private subscriptions: Subscription[] = [];
+  public buttonsBackAndNext: HTMLElement[];
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: this property is not used

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  public sliderIsHovered = false;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: just isHovered

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      {
+        el: this.next.nativeElement,
+        callback: () => this.moveSlide(SLIDE_DIRECTION.RIGHT)
+      },
+      {
+        el: this.back.nativeElement,
+        callback: () => this.moveSlide(SLIDE_DIRECTION.LEFT)
+      }
+    ]);
+    this.firstItem = this.allSlides.nativeElement.querySelectorAll('.slide-item')[0];
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]:
```ts
this.firstSlide = this.allSlides.nativeElement.querySelector('.slide-item:first-child')
this.lastSlide = this.allSlides.nativeElement.querySelector('.slide-item:last-child')
```

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.forEach(subscription => subscription.unsubscribe());
+    clearInterval(this.intervalStart);
+  }
+
+  private moveSlide(direction: SLIDE_DIRECTION): void {
+    this.allSlides.nativeElement.style.transition = '';
+    // default values, that matches LEFT direction to reduce code
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: remove comment

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    let counter = 0;
+    let translatePostion = this.translateStep;
+    let step = this.currentTranslatePosition + this.translateStep;
+    let commonCounterAction = 'dec';
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: I didn't get what is it?

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+      this.commonSlidesMove(translatePostion, commonCounterAction);
+    }
+
+    this.selected = this.counter;
+  }
+
+  private handleClick(arr) {
+    this.subscriptions = arr.map(({ el, callback }) => {
+      const stream = fromEvent(el, 'click');
+      return stream.pipe(throttleTime(400)).subscribe(callback);
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: move throttle time to input property with default value 400

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    });
+  }
+
+  private translateItem(item, step: number): void {
+    if (item.nativeElement) {
+      item.nativeElement.style.transform = `translate(${step}%)`;
+    } else {
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]:
```
const { nativeElement } = item;
if (nativeElement) {
     nativeElement.style.transform = `translate(${step}%)`;
     return;
}
item.style.transform = `translate(${step}%)`;
```

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    this.translateItem(itemToMove, itemMoveDirection);
+    this.currentTranslatePosition = changeTranslatePosition;
+
+    this.allSlides.nativeElement.removeEventListener(
+      'transitionend',
+      this.moveToEdgeSlideWithoutRewind
+    );
+  }
+
+  private handleLastAndFirstMoves(changeSlidesPos, itemToMove, positionToMove, counterAction): void {
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: moveEdgeSlides

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    this.currentTranslatePosition = changeSlidesPos;
+    this.moveContainerAndItem(itemToMove, positionToMove);
+    this.allSlides.nativeElement.addEventListener(
+      'transitionend',
+      this.moveToEdgeSlideWithoutRewind
+    );
+    this.counter = counterAction;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: pass the counter value, not dec/inc string

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  private commonSlidesMove(translatePostion, counterAction): void {
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: specify type of paramenters

commonSlidesMove(translatePostion: number): void

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+      commonCounterAction = 'inc';
+      itemToMove = this.firstItem;
+      itemToMovePosition = this.totalSlidesSize;
+      counterAction = 0;
+    }
+
+    if (this.counter === counter) {
+      this.handleLastAndFirstMoves(step, itemToMove, itemToMovePosition, counterAction);
+    } else {
+      this.commonSlidesMove(translatePostion, commonCounterAction);
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: try to be consistent
rename to moveEdgeSlides and moveSlides

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  public bulletHandler(i: number): void {
+    this.allSlides.nativeElement.style.transition = '';
+    this.counter = i;
+    this.selected = this.counter;
+    this.currentTranslatePosition = i * -this.translateStep;
+    this.translateItem(this.allSlides, this.currentTranslatePosition);
+  }
+
+  public stopSlideshow(): void {
+    this.continueSlideshow = () => null;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: don't overwrite handler. just set a boolean pause: Boolean

and check in the continue function

File: src/app/home-page/components/slideshow/slideshow.component.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.html]: > your

ty for explained comments, i'll fix it
have nice weekend!

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 14 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: ty for explained comments, i'll fix it
have nice weekend!

Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 13 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 12 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 11 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 10 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 9 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 8 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 6 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 5 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 4 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 3 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts (hunk 2 above)
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    let counterAction = this.slidesLength - 1;
+
+    if (direction === SLIDE_DIRECTION.RIGHT) {
+      counter = this.slidesLength - 1;
+      translatePostion = -this.translateStep;
+      step = -100;
+      commonCounterAction = 'inc';
+      itemToMove = this.firstItem;
+      itemToMovePosition = this.totalSlidesSize;
+      counterAction = 0;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: why the var is called Action? the value is number

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+      'transitionend',
+      this.moveToEdgeSlideWithoutRewind
+    );
+    this.counter = counterValue;
+  }
+
+  private moveSlides(translatePostion: number, counterAction: string): void {
+    this.currentTranslatePosition += translatePostion;
+    this.translateItem(this.slidesHolder, this.currentTranslatePosition);
+    counterAction === 'inc' ? this.counter++ : this.counter--;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: pass here the counterValue not counterAction

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    let commonCounterAction = 'dec';
+    let itemToMove = this.lastItem;
+    let itemToMovePosition = -this.totalSlidesSize;
+    let counterAction = this.slidesLength - 1;
+
+    if (direction === SLIDE_DIRECTION.RIGHT) {
+      counter = this.slidesLength - 1;
+      translatePostion = -this.translateStep;
+      step = -100;
+      commonCounterAction = 'inc';
```
[the remark this answers is not in the document]
Comment (taras1313) [reply src/app/home-page/components/slideshow/slideshow.component.ts]: done

File: src/app/home-page/components/slideshow/slideshow.component.ts [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+      itemToMove = this.firstItem;
+      itemToMovePosition = this.totalSlidesSize;
+      counterValue = 0;
+    }
+
+    if (this.currentSlideIndex === edgeSlideIndex) {
+      this.currentSlideIndex = counterValue;
+      this.moveEdgeSlides(step, itemToMove, itemToMovePosition);
+    } else {
+      counterValue ? this.currentSlideIndex-- : this.currentSlideIndex++;
```
Comment (dverbovyi) [inline src/app/home-page/components/slideshow/slideshow.component.ts]: @taras1313  can you use here edgeSlideIndex instead counterValue?

**Pull request merged**